### ISSUE TITLE
[12.0] Wrong context in form views if coming from a tree view.

### ISSUE THREAD

**Issue by LoisRForgeFlow** [odoo/odoo#47680]
Impacted versions:
12.0 (it is fine in v13)

Steps to reproduce:
1. Go to a tree view.
2. Click on a record to navigate to its form view.
3. Call an action.
4. Check context.

Current behavior:
```python
>> self.env.context["params"]["view_type"]
>> "list"

>> self.env.context["params"].get("id")
>> None
```

Expected behavior:
```python
>> self.env.context["params"]["view_type"]
>> "form"

>> self.env.context["params"].get("id")
>> 7
```

Video/Screenshot link (optional):
I did a small mock action to highlight the issue:
```python
class Sales(models.Model):
    _inherit = "sale.order"

    @api.multi
    def action_check_context(self):
        rec_id = self.env.context.get("params", {}).get("id", False)
        model = self.env.context.get("params", {}).get("model", False)
        if not model or not rec_id or model != "sale.order":
            raise UserError("Context is wrong?")
        raise UserError("Everything Good. ID: %s" % rec_id)

```
In the following gif you can see the problem, when coming from the tree view the context is wrong, if I refresh the page, everything seems fine:
![Peek 2020-03-16 11-34](https://user-images.githubusercontent.com/23449160/76748401-fa3a0980-677a-11ea-9464-5d5ae4a2e555.gif)

cc @ForgeFlow

**Comment by pedrobaeza** [comment]
For me this is a misuse of the framework. This is not documented in any place that you can use such internal information for your purpose, and it's by chance that you get it that way on v13. You should find another way of doing what you want to do.

If @Yenthe666 agrees me, we can close this.

**Comment by LoisRForgeFlow** [comment]
@pedrobaeza So how I'm supposed to get the active_id then?

**Comment by Yenthe666** [comment]
Hmmm I'm a bit split about this one. It's not really documented but I'd expect it to always give back the ID too. Maybe @mart-e has an opinion about this one. :)

**Comment by pedrobaeza** [comment]
Define what do you mean by `active_id`. `self` is usually the current record (virtual or already populated in DB). It depends on what do you want to achieve.

**Comment by LoisRForgeFlow** [comment]
@pedrobaeza I want to get the current document ID (in this example sale.order) when I click a button in the a sale order line tree view inside that form view (and please, imagine you don't have the order_id field in the line)

**Comment by pedrobaeza** [comment]
If you have a one2many tree, you must always have the `inverse_name` of that one2many, so there's no need of such artifact to get it.

**Comment by LoisRForgeFlow** [comment]
@pedrobaeza Ok, let's be more specific... I have a computed m2m field in a form view of a model not directly related with it, that m2m is displayed as a tree view inside the form view and it has a button in each line. When I click that button I need to know from which document this action was triggered.

**Comment by pedrobaeza** [comment]
For that m2m field definition: use something like:
```xml
<field name="m2m_field" context="{'whatever_key': id}"/>
```
and then recover from the button method `whatever_key` from the context.

**Comment by LoisRForgeFlow** [comment]
@pedrobaeza That would work in v13, but not in v12 (tried that also)

**Comment by LoisRForgeFlow** [comment]
Extending my answer:

```xml
<field name="computed_m2m_ids" nolabel="1" colspan="2" context="{'test_id_not_passed': id}">
    <tree>
      <field name="name"/>
      <button string="Test action" name="action_test" icon="fa-arrow-up" type="object" 
                  context="{'test_id_passed_but_id_is_m2m_model_not_current_doc': active_id, 'test_id': id}"/>
    </tree>
</field>
```

**Comment by pedrobaeza** [comment]
Have you added explicitly `<id>` field? Anyways, that can be the issue then: the context is not properly propagated.

**Comment by LoisRForgeFlow** [comment]
Just tried, still taking the id of the m2m model.

Thanks for accepting the issue might exist :)

**Comment by mart-e** [comment]
> For me this is a misuse of the framework. This is not documented in any place that you can use such internal information for your purpose, and it's by chance that you get it that way on v13. You should find another way of doing what you want to do.

I agree, the context ids content is not something I would rely on as non-documentated information. But them as this is more a webclient thing, I would try with @aab-odoo maybe.

**Comment by kos94ok-3D** [comment]
What will be if use active_ids[0] instead of active_id?

**Comment by LoisRForgeFlow** [comment]
Don't have neither active_id or active_ids[0], and even if I force them to be passed they refer to the comodel of m2m fields not to the current id in the user's screen.

![image-1](https://user-images.githubusercontent.com/23449160/76770020-907f2700-679d-11ea-89b6-731174005abd.png)

**Comment by LoisRForgeFlow** [comment]
@aab-odoo Any word on this?

Thanks!

**Comment by LoisRForgeFlow** [comment]
ping @aab-odoo I would really appreciate some feedback here

**Comment by LoisRForgeFlow** [comment]
Hi @aab-odoo, it would be great to borrow some of your time to drop a work here :)
